### PR TITLE
RPC and ZeroMQ APIs to support p2pool

### PR DESCRIPTION
**Pull request by SChernykh**
Adds the following:

- "get_miner_data" to RPC API
- "json-full-miner_data" to ZeroMQ subscriber contexts

Both provide the necessary data to create a custom block template. They are used by p2pool.

Data provided:

- major fork version
- current height
- previous block id
- RandomX seed hash
- network difficulty
- median block weight
- coins mined by the network so far
- mineable mempool transactions

### CODE REVIEW

File: src/cryptonote_core/blockchain.cpp [hunk 1]
```diff
@@ -5531,6 +5571,33 @@ void Blockchain::cache_block_template(const block &b, const cryptonote::account_
   m_btc_valid = true;
 }
 
+void Blockchain::send_miner_notifications(const crypto::hash &prev_id, uint64_t already_generated_coins)
+{
+  if (m_miner_notifiers.empty())
+    return;
+
+  const uint64_t height = m_db->height();
```
Comment (vtnerd) [inline src/cryptonote_core/blockchain.cpp:5579]: Would be nice if some of these values could be retrieved lazily only when there is a sub to the corresponding endpoint. But this notification is infrequent, so I guess not worth butchering the interface -

Could you add a comment here about lazily retrieving though? Just to document somewhere that the data (usually) gets dropped on the floor. The notifier signature could be updated to `boost::optional<miner_data>(const crypto::hash&, std::uint64_t /* already_generated_coins */, blockchain&, const miner_data*)`, which is a bit more efficient but admittedly funky.

Basically the first notifier that has a subscriber retrieves the data and returns it for the next one, etc.

File: src/cryptonote_core/tx_pool.cpp [hunk 2]
```diff
@@ -913,6 +913,34 @@ namespace cryptonote
...
+  void tx_memory_pool::get_block_template_backlog(std::vector<tx_block_template_backlog_entry>& backlog, bool include_sensitive) const
+  {
+    CRITICAL_REGION_LOCAL(m_transactions_lock);
+    CRITICAL_REGION_LOCAL1(m_blockchain);
+    const relay_category category = include_sensitive ? relay_category::all : relay_category::broadcasted;
+    backlog.reserve(m_blockchain.get_txpool_tx_count(include_sensitive));
+    cryptonote::blobdata tmp_blob;
+    txpool_tx_meta_t tmp_meta;
+    m_blockchain.for_all_txpool_txes([this, &backlog, &tmp_blob, &tmp_meta](const crypto::hash &txid, const txpool_tx_meta_t &meta, const cryptonote::blobdata_ref *bd){
+      tmp_blob = m_blockchain.get_txpool_tx_blob(txid, relay_category::all);
```
Comment (vtnerd) [inline src/cryptonote_core/tx_pool.cpp:925]: If the `false` on line 941 is changed to `true` then this call is unnecessary because `bd` is a pointer to the transaction blob. This should be a lot more efficient.

File: src/rpc/core_rpc_server.cpp [hunk 3]
```diff
@@ -1861,6 +1861,43 @@ namespace cryptonote
...
+      LOG_ERROR("Failed to get miner data");
+      return false;
+    }
+
+    res.tx_backlog.clear();
+    res.tx_backlog.reserve(tx_backlog.size());
+
+    for (const auto& entry : tx_backlog)
+    {
+      res.tx_backlog.emplace_back(COMMAND_RPC_GETMINERDATA::response::tx_backlog_entry{string_tools::pod_to_hex(entry.id), entry.weight, entry.fee});
```
Comment (vtnerd) [inline src/rpc/core_rpc_server.cpp:1890]: Is the only difference between the types a `std::string` to be compliant with the the epee JSON output engine? That is a glaring mistake in the internal epee DOM (not differentiating between strings and bytes).

File: src/rpc/core_rpc_server_commands_defs.h [hunk 4]
```diff
@@ -940,6 +940,56 @@ namespace cryptonote
     typedef epee::misc_utils::struct_init<response_t> response;
   };
 
+  struct COMMAND_RPC_GETMINERDATA
+  {
+    struct request_t: public rpc_request_base
```
Comment (vtnerd) [inline src/rpc/core_rpc_server_commands_defs.h:945]: I don't see a corresponding change to ZMQ-RPC. Add in a later patch, or ? Because I think this requires pool implementations to use both interfaces (at least for the initialization).

**Review by vtnerd — COMMENTED** [review]
This looks basically ready - comment in `tx_pool` is most "critical" (yet can still be cleanup later). Changing a few lines should make retrieving the tx backlog a bit more efficient.

File: src/rpc/zmq_pub.cpp [hunk 5]
```diff
@@ -116,13 +119,30 @@ namespace
...
   {
     const cryptonote::transaction& tx;
+    crypto::hash hash;
+    uint64_t blob_size;
+    uint64_t weight;
+    uint64_t fee;
   };
 
-  void toJsonValue(rapidjson::Writer<epee::byte_stream>& dest, const minimal_chain self)
+  void toJsonValue(rapidjson::Writer<epee::byte_stream>& dest, const minimal_chain& self)
```
[the remark this answers is not in the document]
Comment (SChernykh) [reply src/rpc/zmq_pub.cpp:145]: The `&` was intentionally added back as passing by reference is still cheaper than copying 24 bytes, also to make all `toJsonValue` functions in zmq_pub.cpp follow the same standard.

File: src/rpc/zmq_pub.cpp [hunk 6]
```diff
@@ -116,13 +119,30 @@ namespace
     const epee::span<const cryptonote::block> blocks;
   };
 
+  //! Object for miner data serialization
+  struct miner_data
```
[the remark this answers is not in the document]
Comment (SChernykh) [reply src/rpc/zmq_pub.cpp:123]: I don't see why when other structs are right next to it. Removing `&` will make a copy of the vector which is bad for performance. It is currently created in one place and never copied, only the reference is passed around. This vector can contain hundreds of transactions.

File: src/rpc/core_rpc_server.cpp (hunk 3 above)
Comment (SChernykh) [reply src/rpc/core_rpc_server.cpp:1890]: JSON outputs binary data here for entry.id if I don't convert to string explicitly.

File: src/rpc/core_rpc_server_commands_defs.h (hunk 4 above)
Comment (SChernykh) [reply src/rpc/core_rpc_server_commands_defs.h:945]: ZMQ-RPC doesn't have half of needed APIs - not even get_block_template/submit_block, so pool implementations are forced to use regular RPC.

File: src/cryptonote_core/blockchain.cpp (hunk 1 above)
Comment (SChernykh) [reply src/cryptonote_core/blockchain.cpp:5579]: `m_miner_notifiers` is not empty only when there is `--zmq-pub` in monerod command line, so it early outs in 99% cases. The remaining monerod setups that do use `--zmq-pub` most likely will use it for p2pool, so they will use this data. Refactoring it makes sense later when there are more uses for ZMQ.

Comment (vtnerd) [reply src/cryptonote_core/blockchain.cpp:5579]: Mining is not the only use case for pub/sub, so there will be cases where all of this information is collected only to be dropped. The other endpoints are forwarding data that is already available.

File: src/rpc/zmq_pub.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (vtnerd) [reply src/rpc/zmq_pub.cpp:123]: The `&` would not be a copy because the entire `struct` would be passed by reference. And it would cleanup the long parameter list.

File: src/rpc/core_rpc_server_commands_defs.h (hunk 4 above)
Comment (vtnerd) [reply src/rpc/core_rpc_server_commands_defs.h:945]: Seems like something to add in another PR then.

File: src/cryptonote_core/blockchain.cpp (hunk 1 above)
Comment (SChernykh) [reply src/cryptonote_core/blockchain.cpp:5579]: Is there an easy way to detect if anyone is connected to zmq-pub port and subscribed to `json-full-miner_data`? I tried to do something like this initially, but I couldn't find a way to detect connected subscribers reliably. IIRC ZMQ somehow cached it internally and `add_subscriptions` was only called once even when I connected two subscribers, or something like that.

File: src/rpc/zmq_pub.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (vtnerd) [reply src/rpc/zmq_pub.cpp:145]: I re-checked this - the ASM generated should be the same currently, so might as well make the change.

If the `struct` were just 2 fields, then each field could be passed in a register. The only caller doesn't have a `minimal_chain` object in working memory, so its copying the data anyway, its just a matter of how and where. However, this is caller specific, and there's a clause I missed at the end of the ABI document that restricts when this type of optimization can occur (from 4 fields to just 2 fields).

File: src/cryptonote_core/blockchain.cpp (hunk 1 above)
Comment (vtnerd) [reply src/cryptonote_core/blockchain.cpp:5579]: The `miner_subs_` array should be an accurate count. ZMQ may coalesce TCP connections on the client side, so you need to create two `zmq_context` objects or use a separate process entirely. If this still doesn't work, then I'd have to re-read the documentation because I misunderstood how `XPUB` works.

File: src/cryptonote_core/blockchain.cpp [hunk 7]
```diff
@@ -4362,6 +4392,7 @@ bool Blockchain::handle_block_to_main_chain(const block& bl, const crypto::hash&
   get_difficulty_for_next_block(); // just to cache it
   invalidate_block_template_cache();
 
+  send_miner_notifications(id, m_db->get_block_already_generated_coins(m_db->height() - 1));
```
Comment (j-berman) [inline src/cryptonote_core/blockchain.cpp:4395]: Is it not safe to use the `already_generated_coins` already in memory here?

Comment (SChernykh) [reply src/cryptonote_core/blockchain.cpp:4395]: I actually didn't see it. I'll test it and if it has the same value I'll change the code.

File: src/rpc/zmq_pub.cpp [hunk 8]
```diff
@@ -143,19 +163,27 @@ namespace
...
-    if (!get_transaction_hash(self.tx, id, blob_size))
-    {
-      MERROR("ZMQ/Pub failure: get_transaction_hash");
-      return;
-    }
+    dest.StartObject();
+    INSERT_INTO_JSON_OBJECT(dest, major_version, self.major_version);
+    INSERT_INTO_JSON_OBJECT(dest, height, self.height);
+    INSERT_INTO_JSON_OBJECT(dest, prev_id, self.prev_id);
+    INSERT_INTO_JSON_OBJECT(dest, seed_hash, self.seed_hash);
```
Comment (j-berman) [inline src/rpc/zmq_pub.cpp:172]: Thoughts on treating `seed_hash` as optional if `major_version >= RX_BLOCK_VERSION` (here & in other parts of the code too) just in case the endpoint is being used for earlier blocks too?

Comment (SChernykh) [reply src/rpc/zmq_pub.cpp:172]: It's always used for the current blockchain height, not for any earlier blocks. Even if it's used for major_version before that, seed_hash will just be all 0, but it only matters for testnet experiments (mining older blocks for example).

Comment (j-berman) [reply src/rpc/zmq_pub.cpp:172]: Sounds fine to me :)
